Energenie sockets preserve state

NotBobTheBuilder opened the pull request:
At present when you initialise the `Energenie` class, sockets are either switched on or off according to `initial_value`.

Sometimes it is useful to initialise the class without setting a state, for example starting a service to control a socket which respects its initial state.

This PR allows specifying `None` as an `initial_value`, in which case no message is sent to the socket. 

The default behaviour of switching off the socket remains unchanged

bennuttall commented:
Thanks Jack. I seem to remember @waveform80 intentionally not allowing this, because you can't read the initial state, so if you tried to read the value after init, before explicitly turning it on or off, it might "lie" to you.

A compromise may be to allow `initial_value` to be `None`, but then let `value` be `None` until it's been set by turning the device on or off.

I'll try and get a decision from Dave.

NotBobTheBuilder commented:
Got it. The compromise you described sounds good to me. I can add a docstring noting `value` could be `None` if an `initial_value` of `None` is specified if that's the way we go.

For context I am using this class in a simple daemon to which other processes send on/off messages. This process starts up without knowing whether the socket should be on or off and another process sends the control messages, so it's helpful for it to start up without making any changes to the outside world.

bennuttall commented:
Sounds good to me. Push another commit and I'll try it out.

Will defer to Dave to make the final call, since we wrote it in the first place. But this seems a reasonable way of dealing with it.

bennuttall commented:
Maybe look at adding tests for this edge case if you can.

lurch commented:
It's actually the "other side of the coin" but you may find the discussions in #707 and #323 of interest too :wink: 

I wondered if having `value == None` might be a weird thing for the GpioZero API to support, but then I remembered that we use it in servo https://gpiozero.readthedocs.io/en/stable/api_output.html#gpiozero.Servo.value
"This can also be the special value None indicating that the servo is currently “uncontrolled”, i.e. that no control signal is being sent."

So I guess it might make sense that for Energenie "This can also be the special value None indicating that the state of the socket is unknown"? And obviously while you could read a `None` value from an Energenie, you should be prevented from setting `.value` to None.

lurch commented on gpiozero/boards.py:
```diff
@@ -2045,11 +2045,13 @@ class Energenie(SourceMixin, Device):
+        means of reading their state, you may provide an initial state for
         the socket, which will be set upon construction. This defaults to
         :data:`False` which will switch the socket off.
+        Specifying :data:`None` will not switch the socket on construction.
```
"will not alter the socket's state" might be less ambiguous than "will not switch the socket on construction" ?

("not switch the socket on" could be misinterpreted as "switch the socket off" ? :wink: )

lurch commented on gpiozero/boards.py:
```diff
@@ -2063,15 +2065,15 @@ def __init__(self, socket=None, initial_value=False, pin_factory=None):
         if initial_value:
             self.on()
-        else:
+        elif initial_value is False:
```
I don't think we should be checking explicitly for True or False, but instead simply checking for None, and then just using duck-typing otherwise? See e.g. https://github.com/gpiozero/gpiozero/blob/master/gpiozero/input_devices.py#L112

I think this also means the `if initial_value not in (True, False, None):` line would need to be removed?

NotBobTheBuilder commented:
Thanks for all the feedback! Now updated. Would be helpful if someone could double check the `pins` part of the tests I added, as I am not 100% sure how the internals of that section works. Newly added test sections are passing, though.

lurch commented on gpiozero/boards.py:
```diff
@@ -2102,6 +2105,9 @@ def value(self):
         """
         Returns :data:`True` if the socket is on and :data:`False` if the
         socket is off.  Setting this property changes the state of the socket.
+        Returns :data:`None` only when constructedwith :data:`initial_value`
```
Missing space in "constructedwith" :wink:

lurch commented on gpiozero/boards.py:
```diff
@@ -2102,6 +2105,9 @@ def value(self):
         socket is off.  Setting this property changes the state of the socket.
+        Returns :data:`None` only when constructedwith :data:`initial_value`
+        set to :data:`None` and neither :data:`on()` nor :data:`off()` have
+        been called since construction
```
Missing full-stop at end of sentence.

lurch commented on gpiozero/boards.py:
```diff
@@ -2063,13 +2066,13 @@ def __init__(self, socket=None, initial_value=False, pin_factory=None):
         self._master = _EnergenieMaster(pin_factory=pin_factory)
-        if initial_value:
+        if initial_value is None:
+            pass
```
Ugh :slightly_frowning_face: 
Could you please make this:
```python3
if initial_value is not None:
    if initial_value:
        self.on()
    else:
        self.off()
```

lurch commented on tests/test_boards.py:
```diff
@@ -1295,6 +1297,15 @@ def test_energenie(mock_factory):
+        pins[2].assert_states_and_times([(0.0, True), (0.0, False)])
+        pins[3].assert_states_and_times([(0.0, True), (0.0, False)])
+        pins[4].assert_states_and_times([(0.0, False)])
+        pins[5].assert_states_and_times([(0.0, False), (0.1, True), (0.25, False)])
```
Would you mind adding tests that `device3.value = 0` and `device3.value = 1` work as expected, and `device3.value = None` raises an exception? Thanks!

lurch commented on gpiozero/boards.py:
```diff
@@ -2063,13 +2066,13 @@ def __init__(self, socket=None, initial_value=False, pin_factory=None):
         if not (1 <= socket <= 4):
             raise EnergenieBadSocket('socket number must be between 1 and 4')
-        if initial_value is None:
-            raise EnergenieBadInitialValue("initial value can't be None")
```
Just for the sake of tidiness, perhaps it's worth removing this exception from the library? :shrug:

lurch replied:
...and also from `exc.py` ? (sorry for not being more explicit)

NotBobTheBuilder replied:
Should be removed :) (maybe you are seeing a partial diff?)

lurch commented on gpiozero/boards.py:
```diff
@@ -2102,11 +2102,16 @@ def value(self):
     @value.setter
     def value(self, value):
+        if value is None:
+            raise TypeError('value cannot be None')
```
A more specific exception would be better - hmm perhaps `EnergenieBadInitialValue` in `exc.py` could be renamed to `EnergenieBadValue` (with the detail-string updated too) and then used here? :slightly_smiling_face:

lurch commented on tests/test_boards.py:
```diff
@@ -1295,6 +1297,25 @@ def test_energenie(mock_factory):
+        device3.value = True
+        try:
+            device3.value = None
+        except:
```
Should probably use the `with pytest.raises(ValueError)` syntax rather than using try / except?

lurch commented on tests/test_boards.py:
```diff
@@ -1295,6 +1297,25 @@ def test_energenie(mock_factory):
+        pins[5].assert_states_and_times([(0.0, False), (0.1, True), (0.25, False)])
+
+        device3.value = False
+        device3.value = True
```
I'd probably make this:
```python3
        device3.value = False
        assert not device3.value
        device3.value = True
        assert device3.value
```
(i.e. check that the value you set is also the value you read back!)

lurch commented on gpiozero/boards.py:
```diff
@@ -2063,15 +2065,13 @@ def __init__(self, socket=None, initial_value=False, pin_factory=None):
         if initial_value:
             self.on()
-        else:
+        elif initial_value is not None:
```
Sorry to be such a nitpicker, but although what you're proposing here is "more compact", I still prefer the "more explicit" version that I suggested:
```python3
if initial_value is not None:
    if initial_value:
        self.on()
    else:
        self.off()
```

I guess it's up to @bennuttall really though!

bennuttall commented:
Dave approves!
